event_routing_backends/backends/events_router.py

```diff
@@ -154,8 +163,27 @@ def send(self, event):
         Arguments:
             event (dict): the original event dictionary
         """
-        event_routes = self.prepare_to_send([event])
+        if settings.EVENT_ROUTING_BACKEND_BATCHING_ENABLED:
+            redis = get_redis_connection()
+            batch = self.queue_event(redis, event)
+            if not batch:
+                return
+
+            try:
+                redis.set(self.last_sent_key, datetime.now().isoformat())
+                self.bulk_send([json.loads(queued_event.decode('utf-8')) for queued_event in batch])
+            except Exception:  # pylint: disable=broad-except
+                logger.exception(
+                    'Exception occurred while trying to bulk dispatch {} events.'.format(
+                        len(batch)
+                    ),
+                    exc_info=True
+                )
+                logger.info('Re sending the batched events to the queue.')
+                redis.lpush(self.queue_name, *batch)
```

Comment:
Thinking about this some more, I think we should have a "dead letter queue" instead. A separate queue that only stores failed batches. This would be for the case when a malformed event gets rejected by the LRS, which would otherwise cause the batch to be reprocessed endlessly, while the queue continues to grow.

We can monitor and alarm on the dead letter queue, and add a management command to move events from it back to the main queue when things are working again. I think we would probably also need to add some kind of filter to remove actually invalid events when doing that move, that takes... maybe an event id, verb id, or object id and discards matching events?

Reply:
When a batch fails and needs to be reprocessed could we make it so that it is not processed as batch any more, but instead it is re-processed as single events?
Maybe this could be done through the "dead letter queue" or splitting the batch here.

Reply:
I was thinking of the same mechanism using split batches, by cutting it in half till it finishes but the problem is that when Ralph is down it adds an unnecessary amount of processing to events that will timeout or will be lost at the end.

But now that you say it this way, we can send the events to the dead letter queue (e.g: `FAILED_ROUTING_EVENT`) and send events one by one via a Django command once operators confirm Ralph is up again. `tutor-contrib-aspects` can provide a job for it.

In this case, we would also need to add some kind of monitoring in Superset so that the operators will know when to jun the job

Reply:
Yeah I like the "send to dead letter, but re-process individually" plan. 👍 We can add dead letter monitoring to the thing I'm writing now once both are implemented the way we want them.

Reply:
We probably want to queue them individually to the dead letter queue so they can be processed in a way that's guaranteed to not break the whole batch on re-processing.